scripts/source-maintenance.sh

```diff
@@ -203,8 +203,105 @@ checkMakeNamedErrorDetails ()
     return $problems
 }
 
+# extract IDs and gists of cache_log_message debugs() in the given source file
+collectDebugMessagesFrom ()
+{
+    source="$1"
+    destination="doc/debug-messages.tmp"
+
+    # Merge multi-line debugs() into one-liners and remove '//...' comments.
+    awk 'BEGIN { found=0; dbgLine=""; } {
+        if ($0 ~ /[ \t]debugs[ \t]*\(/)
+            found = 1;
+        if (found) {
+            commented = match($0, /\);[ \t]*\/\//);
+            if (commented)
+                $0 = substr($0, 1, RSTART+1);
+            dbgLine = dbgLine $0;
+        }
+        if ($0 ~ /\);/) {
+            if (found) {
+                found = 0;
+                print dbgLine;
+                dbgLine = "";
+            }
+        }
+    }' $source > doc/debug-messages.tmp2
+
+    # sed expressions:
+    # - replace debugs() prefix with the message ID contained in it
+    # - remove simple parenthesized non-"string" items like (a ? b : c)
+    # - replace any remaining non-"string" items with ...
+    # - remove quotes around "strings"
+    # - remove excessive whitespace
+    # - remove debugs() statement termination sugar
+    grep -o -E '\bdebugs[^,]*,\s*(Critical|Important)[(][0-9]+.*' doc/debug-messages.tmp2 | \
+        sed -r \
+            -e 's/.*?(Critical|Important)[(]([0-9]+)[)],\s*/\2 /' \
+            -e 's/<<\s*[(].*[)]\s*(<<|[)];)/<< ... \1/g' \
+            -e 's/<<\s*[^"]*/.../g' \
+            -e 's@([^\\])"@\1@g' \
+            -e 's/\s\s*/ /g' \
+            -e 's/[)];$//g' \
+        >> $destination
+
+    rm -f doc/debug-messages.tmp2
+}
+
+# make doc/debug-messages.dox from aggregate collectDebugMessagesFrom results
+processDebugMessages ()
+{
+    source="doc/debug-messages.tmp"
+    destination="doc/debug-messages.dox"
+
+    repeatedIds=`awk '{print $1}' $source | sort -n | uniq -d`
+    if test "x$repeatedIds" != "x"; then
+        echo "ERROR: Repeated debugs() message IDs:"
+        echo "$repeatedIds"
+        echo ""
+        return 1;
+    fi
+
+    repeatedGists=`awk '{$1=""; print substr($0,2)}' $source | sort | uniq -d`
+    if test "x$repeatedGists" != "x"; then
+        echo "ERROR: Repeated debugs() message gists:"
+        echo "$repeatedGists"
+        echo ""
+        return 1;
+    fi
+
```

Comment:
We now [know](https://github.com/squid-cache/squid/pull/775/commits/e784de0a9fd97e47adf84d5cbe7e49a90fff82c7) that basic [awk] compatilibity can result in no IDed debugs() messages found in Squid sources. If it is not too difficult, please add a third check, validating that `$source` is not empty (i.e. that at least one IDed debugs() message was found). I do not think we should _count_ the number of found messages (for now) because there is currently no good constant to compare that counter with. Just a simple `test '!' -s "$source` will be sufficient for now.

I do not insist on this (somewhat paranoid) change. The new check cannot detect many message-finding problems.

Reply:
Done (aaedc72).